### ISSUE TITLE
how to restore an a big size arangodump

### ISSUE THREAD

**Issue by zeroluojin** [arangodb/arangodb#8162]
## My Environment

* __ArangoDB Version__:        3.4.2
* __Storage Engine__:             RocksDB 
* __Deployment Mode__:       
* __Deployment Strategy__:   Manual Start
* __Infrastructure__:               physical server/ PC server
* __Operating System__:       centos 7 
* __Total RAM in your machine__:       e.g. 256G
* __Disks in use__:      SSD 
* __Used Package__:              

__Problem__:
I have a very huge collection, almost 300 million documents. when i used arangodump and arangorestore command to restore other server (two server's setting is the same),  restore performance is too slow and always failed I don't know something is wrong. and then I try to upgrade to 3.4 version . and add batch size and threads parameter. but still unable to restore.
below is my backup and restore command.  

```
nohup arangodump --server.endpoint tcp://10.102.1.101:8529 --server.username root --server.password Password01! --server.database kgms  --output-directory "/opt/dump_prod_2019_1_30" 2>&1 &
```

```
nohup arangorestore --server.endpoint tcp://10.102.1.104:8529 --server.username root --server.password Password01! --server.database kgms --create-database true --collection kg_entity_12236201 --collection kg_edage_12236201 --input-directory "dump_prod_2019_1_30"  --threads 4  --batch-size 100663296 --force true 2>&1 &
```

anyone can give me some suggestions, thanks a lot.
![arango](https://user-images.githubusercontent.com/43881415/52768608-f401c700-3068-11e9-9326-c4d202c168cd.jpg)

**Comment by Simran-B** [comment]
Can you add the details about your environment please?

**Comment by zeroluojin** [comment]
ArangoDB Version: 3.4.2
Storage Engine: rocksdb
Deployment Mode: manual set up
Infrastructure: physical server/ PC server
Operating System: centos 7
Total RAM in your machine: 256 G
Disks in use: SSD

**Comment by OmarAyo** [comment]
Hi @zeroluojin,

The value passed for the option `--batch-size` seems to be to high (100MB), can you lower that value and retry ? We suggest the default value of ` 8388608` 

Best,

**Comment by zeroluojin** [comment]
> Hi @zeroluojin,
> 
> The value passed for the option `--batch-size` seems to be too high (100MB), can you lower that value and retry ? We suggest the default value of ` 8388608`
> 
> Best,

thank you for you reply 
Actually, for this option value, i have many times experience, things like default value of  8388608,
50 MB, 80 MB, 100 MB. but if i set 8388608, the performance is too slow,  the progress is 20% need to take 2 days and the last information is the response time out. 

and I have another question, can i just direct to copy data file directory to another server. to implement 'cold backup' ?

**Comment by dothebart** [comment]
Closing this as [hot backup](https://www.arangodb.com/docs/stable/backup-restore.html) is available. 

In other news, [ArangoDB 3.7](https://www.arangodb.com/2020/08/arangodb-3-7-a-big-step-forward-for-multi-model/) is available.

**Comment by admtech** [comment]
Warum wurde Issue geschlossen? Das Problem besteht weiterhin. Auch ich habe die gleichen Probleme mit Arangorestore. Arangodump funktioniert mittlerweile super schnell, Arangorestore ist weiter super langsam. Hot Backup ist sicher eine Alternative, aber nur wer die Enterprise Version benutzt. Es sollte auch eine vernünftige Alternative ohne Enterprise geben. Andere DB bekommen das auch hin. Auch ArangoDb 3.7 löst das Problem nicht. Ich habe eine 2,4 GB große Datenbank und sie braucht für den Dump 70 Sekunden, für den Import aber über 22 Minuten!! Hardware: 12 Kerne, 32 GB RAM, SSD mit 3000MB/sec. Alle --thread und --batch-size Einstellungen wurden bereits ausprobiert (8,80,800 MB). Der Import variiert nur in 1 bis zwei Minuten, bleibt aber immer über 20 Minuten. Der Issue sollte wieder geöffnet werden, da Arangorestore im Moment als Backup bzw. Restore nicht zufriedenstellend funktioniert und kein funktionierendes Backup-Konzept dafür ist. Danke

**Comment by dothebart** [comment]
Long story short: restoring requires much more computing power than dumping, hence this can not become be as fast. 
Detail reasons:
- indices have to be maintained; hence hashes have to be calculated, and checked for clashes
- documents have to be compressed
- RocksDB will do internal compaction runs to ensure disk usage is acceptable
- replication has to send documents to followers 

At least the RocksDB step could be optimized, by configuring your environment along the lines to become faster:
[Reducing the Memory Footprint of ArangoDB servers](https://www.arangodb.com/docs/stable/tutorials-reduce-memory-footprint.html)
(while the title the article will make you expect the opposite, the article explains which screws to turn to reduce the footprint and what the resulting effects are. For sure these effects can also be moved into the opposite direction using the same paramters)

HotBackup on the other hand uses orchestrated approach to get RocksDB snapshots, and to be able to get in the shortest possible time frame a persistent on disk state, which it can use to copy the data afterward this snapshotting has been finished, while the actual operations may continue.

 But, as you already said, its an enterprise feature, and it will remain that way. So from our perspective this issue *is* solved. 

The next best approach you could choose is to do a short stop of your database, and use LLVM to capture a snapshot of your on disk data.

**Comment by admtech** [comment]
> But, as you already said, its an enterprise feature, and it will remain that way. So from our perspective this issue is solved.

Really, that's your answer to a problem. We have a premium feature, so that's the end of the matter for us? Is this a typical German answer? I find this very clumsy, especially for developers who want to switch to ArangoDB.

A reasonable import/export should offer every database without enterprise version. Of course I understand that an import always takes longer, I have been working with databases for over 20 years. But with you the time ratio of export/import is just not right: 70 sec vs 1200 sec ?? I have already considered the article "Reducing the Memory Footprint ...". We run a website with ArangoDB, so we can't just stop the database.

Maybe you should think about it again and get a developer to do it. I think I am not the only developer who would like to use ArangoDB, but finds the import time clearly too long.

**Comment by fceller** [comment]
We should not close this issue unless we understand why the restore is slow.

**Comment by admtech** [comment]
Thanks

**Comment by jsteemann** [comment]
@admtech : I think I am a bit late to the game here, so just trying to understand the issue.
Is my following understanding correct?

- the current ArangoDB version in use is 3.7.x, but the problem was always present (i.e. no new issue)
- the performance of arangodump is fine, but the performance of arangorestore is not (restore is about factor 20 slower than dump)
- the restore is done into a single server (i.e. no cluster)
- when restoring, the target database/collection do not exist
- there is one large collection that is restored and that is creating the problem. if not, how many collections are restored and what are their sizes?

Especially the last point would be important to understand. If there is more than one collection to restore, arangorestore can use multiple threads (controlled via `--threads` option) to restore the data into multiple collections concurrently.
However, if there is one collection which is substantially larger than the other collections in the dump, the restore may work effectively single-threaded for a long time while this collection is processed.
If the restore is really working mostly on a single collection, I am wondering if splitting the input file into multiple chunks would help here, and then parallelizing the restore of the chunks. That is nothing that arangorestore can do right now, but it may be a way to improve it for large collections.

What may also have an impact on restore performance is the collection's `waitForSync` property, whether or not the collection has a schema attached and the secondary indexes that are present for the collection. Can you share these details with us?
From the arangosh, that would be the output of `db.<collection-name>.properties();` and of `db.<collection-name>.indexes()`. 

Finally, starting in ArangoDB 3.8 there is a new option for arangodump (`--envelope`) that controls the dump format. The default value for it is `true`, meaning the format will remain unchanged. However, when setting this option to `false`, the generated dumps will be smaller and slightly quicker to produce. arangorestore in version 3.8 can handle such dumps automatically, and I guess the restore time will also go down by using this format.
I don't expect anything near a factor of 20 here, rather a small percentage.

To make restore massively faster we may need to parallelize restoring of a single collection. But it would be good if you could let me know if my above assumptions are appropriate, and if you could share the collection properties and indexes info. 
Thanks!

**Comment by admtech** [comment]
Hello @jsteemann,

I'm glad you're following up. I'll try to answer all the questions:

> the current ArangoDB version in use is 3.7.x, but the problem was always present (i.e. no new issue)

yes, ArangoDB version 3.7.10 and the problem has existed since we started using Arangodb (we started with 3.4)

> the performance of arangodump is fine, but the performance of arangorestore is not (restore is about factor 20 slower than dump)

Exactly, the speed of arangodump is very good.  Our database is 2.4 GB and it takes about 70 seconds. For the restore of the 2.4 GB database it currently needs 20 to 22 minutes (AMD 12 cores, 32 GB RAM, SSD with 3000MB/sec.) So it can't be the performance of the hardware or disks.

> the restore is done into a single server (i.e. no cluster)

yes, currently it is still a single server

> when restoring, the target database/collection do not exist

Good question: we use: "arangorestore --create-database" and for the test we imported all this into a new DB. It didn't exist the first time, but it did for all subsequent tests. I'll test if there is a difference in the import. In any case, it is not a DB that anyone else is accessing in parallel.

> there is one large collection that is restored and that is creating the problem. if not, how many collections are restored and what are their sizes?

Our database contains several large table:

- Articles = 500.000 entries
- Comments = 1.5 million entries
- Jobs =. 60.000 entries
- Users = 130.000 entries
- Messaging system =. 200.000 entries
- Geo data = 300.000 entries
- and then many small collections

Articles, Users and Comments are the biggest items, also in terms of volume. It takes the longest time for these 3 collections.

> Especially the last point would be important to understand. If there is more than one collection to restore, arangorestore can use multiple threads (controlled via --threads option) to restore the data into multiple collections concurrently.

We have tested the "-threads x" option extensively (4, 6, 8, 12). With "dump" it really brings a lot, with "restore" it has almost no effect with us.

> However, if there is one collection which is substantially larger than the other collections in the dump, the restore may work effectively single-threaded for a long time while this collection is processed.

These are then the above described three collection "User", "Comments" and "Articles". A certain parallelism by "threads" I can not recognize on the basis of the output, however.

> If the restore is really working mostly on a single collection, I am wondering if splitting the input file into multiple chunks would help here, and then parallelizing the restore of the chunks. That is nothing that arangorestore can do right now, but it may be a way to improve it for large collections.

But then arangodump would have to support that as well.

> What may also have an impact on restore performance is the collection's waitForSync property, whether or not the collection has a schema attached and the secondary indexes that are present for the collection. Can you share these details with us?
From the arangosh, that would be the output of db.<collection-name>.properties(); and of db.<collection-name>.indexes().

Schemas are not attached. The "waitForSync" property is set to "false" for all collections.

Here are the outputs of the big 3 collections:

COMMENTS:

db.comments.properties()
{
  "isSystem" : false,
  "waitForSync" : false,
  "keyOptions" : {
    "allowUserKeys" : true,
    "type" : "traditional",
    "lastValue" : 67350216
  },
  "writeConcern" : 1,
  "cacheEnabled" : false,
  "schema" : null
}

 db.comments.indexes()
[
  {
    "fields" : [
      "_key"
    ],
    "id" : "comments/0",
    "name" : "primary",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "primary",
    "unique" : true
  },
  {
    "deduplicate" : true,
    "fields" : [
      "status",
      "languages",
      "stamp_create"
    ],
    "id" : "comments/82766853",
    "name" : "stamp_create",
    "selectivityEstimate" : 0.9974430780469986,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "_key",
      "status"
    ],
    "id" : "comments/82766904",
    "name" : "comment_key",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "persistent",
    "unique" : true
  },
  {
    "deduplicate" : true,
    "fields" : [
      "rating.count",
      "rating.user",
      "status"
    ],
    "id" : "comments/82766924",
    "name" : "rating",
    "selectivityEstimate" : 0.00001980741533037354,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "reply_to_comment",
      "status"
    ],
    "id" : "comments/82766934",
    "name" : "reply",
    "selectivityEstimate" : 0.21391268017547524,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "content_key",
      "status"
    ],
    "id" : "comments/82766944",
    "name" : "content_key",
    "selectivityEstimate" : 0.13676811860360286,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "content_key",
      "status",
      "reply_to_comment"
    ],
    "id" : "comments/82766958",
    "name" : "detail_comments",
    "selectivityEstimate" : 0.3000952450728991,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  }
]

USER:

db.users.properties()
{
  "isSystem" : false,
  "waitForSync" : false,
  "keyOptions" : {
    "allowUserKeys" : true,
    "type" : "traditional",
    "lastValue" : 70904606
  },
  "writeConcern" : 1,
  "cacheEnabled" : false,
  "schema" : null
}

db.users.indexes()
[
  {
    "fields" : [
      "_key"
    ],
    "id" : "users/0",
    "name" : "primary",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "primary",
    "unique" : true
  },
  {
    "deduplicate" : true,
    "fields" : [
      "status",
      "stamp_create",
      "language"
    ],
    "id" : "users/83718800",
    "name" : "create",
    "selectivityEstimate" : 0.9915274751876059,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "user_hash",
      "status",
      "personal.privacy"
    ],
    "id" : "users/83718804",
    "name" : "user_profile",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  }
]

CONTENT:

db.content.properties()
{
  "isSystem" : false,
  "waitForSync" : false,
  "keyOptions" : {
    "allowUserKeys" : true,
    "type" : "traditional",
    "lastValue" : 78444052
  },
  "writeConcern" : 1,
  "cacheEnabled" : false,
  "schema" : null
}

db.content.indexes()
[
  {
    "fields" : [
      "_key"
    ],
    "id" : "content/0",
    "name" : "primary",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "primary",
    "unique" : true
  },
  {
    "deduplicate" : true,
    "fields" : [
      "user_key",
      "languages",
      "status"
    ],
    "id" : "content/87160379",
    "name" : "user_key",
    "selectivityEstimate" : 0.28707597420801795,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "status",
      "primary_language"
    ],
    "id" : "content/87160383",
    "name" : "primary_language",
    "selectivityEstimate" : 0.000007567273057481006,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "tag_key",
      "languages",
      "status"
    ],
    "id" : "content/87160387",
    "name" : "tag_keys",
    "selectivityEstimate" : 0.000609165481127221,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "art",
      "status",
      "tag_key",
      "tags"
    ],
    "id" : "content/87160391",
    "name" : "content_counter",
    "selectivityEstimate" : 0.002746920119865605,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "tags",
      "status",
      "art",
      "languages",
      "stamp_create"
    ],
    "id" : "content/87160401",
    "name" : "content_tags_create",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "status",
      "languages",
      "stamp_create"
    ],
    "id" : "content/87160409",
    "name" : "create_date",
    "selectivityEstimate" : 0.9997559189650964,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "statistic.comments_count",
      "status"
    ],
    "id" : "content/87160423",
    "name" : "comment_counter",
    "selectivityEstimate" : 0.000003783636528740503,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "stamp_last_activity",
      "status"
    ],
    "id" : "content/87160427",
    "name" : "last_activities",
    "selectivityEstimate" : 0.999633923123856,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "status",
      "statistic.read_count",
      "stamp_create",
      "languages"
    ],
    "id" : "content/87160437",
    "name" : "create_status_read_count",
    "selectivityEstimate" : 1,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  },
  {
    "deduplicate" : true,
    "fields" : [
      "status",
      "art",
      "stamp_create",
      "languages"
    ],
    "id" : "content/87160449",
    "name" : "side_content",
    "selectivityEstimate" : 0.9995119570522206,
    "sparse" : false,
    "type" : "persistent",
    "unique" : false
  }
]

> Finally, starting in ArangoDB 3.8 there is a new option for arangodump (--envelope) that controls the dump format. The default value for it is true, meaning the format will remain unchanged. However, when setting this option to false, the generated dumps will be smaller and slightly quicker to produce. arangorestore in version 3.8 can handle such dumps automatically, and I guess the restore time will also go down by using this format.
I don't expect anything near a factor of 20 here, rather a small percentage.

Sounds good and we will give it a try.

> To make restore massively faster we may need to parallelize restoring of a single collection. But it would be good if you could let me know if my above assumptions are appropriate, and if you could share the collection properties and indexes info.
Thanks!

I am happy to do so at any time.

Greetings
Frank

**Comment by jsteemann** [comment]
@admtech : I spent several hours looking into the reasons for the slow restore performance.
To simplify things for me, I focused on a single use case: restoring one large collection from a dump, with no secondary indexes.
Although this is not exactly your use case, improving this particular case should improve a lot of other cases too, e.g. restoring multiple large collections or restoring a mix of large and small collections.

One (known) problem with arangorestore is that when restoring only a single collection, it can't use any multi-threading. The restore thread that reads the input file blocks when sending a batch of data to the server until a response is received. Thus arangorestore is idle while a request is in flight (of course, this is only true when restoring a single collection - when restoring multiple collections the other threads can still do other work).

I have hammered out a change to arangorestore that can overcome this limitation, so that the thread that is reading the input file can dispatch the request-response handling to another thread. This means we will have the other thread working and the original thread can go on and read more of the input file: more parallel work and less idle time on the arangorestore side!
The restore thread that reads the input file is supposed to read the input data faster than what arangorestore can get to the server via request-response. In order to maximize parallel work, the reading thread is allowed to dispatch as many request-response jobs as there are idle arangorestore threads.
For example, when restoring a single collection with `--threads 8`, the thread reading the input file can dispatch to up to 7 other threads, so that up to 8 concurrent requests can be in flight for one collection.

While this sounds nice, there are also a few caveats:
- for now, the change is only available in a PR: https://github.com/arangodb/arangodb/pull/14010. If the change works fine, it will be merged into devel, which will become 3.9 eventually. It may be backported to 3.8 as well, but very likely not to 3.7.
- increased parallelism can only be used when restoring dumps that were created with the non-enveloped format (i.e. an arangodump run with `--envelope false` option). This option is available from 3.8 onwards, and its defaults to `true`. In 3.8 this option is opt-in. In devel/3.9 the PR will change the default value to `false`, so dumps will be slick and restore can use the improved parallelism
- increasing the number of arangorestore threads currently doesn't lead to a linear speedup of the restore process. Rather, the speedup is sub-linear. This is because the work can be bottlenecked on reading and decompressing the input file. In addition, there is a critical section without multi-threading when a chunk of data is handed over to another thread and needs to be copied for that. Finally, restoring data will only be as fast as the server can ingest the data. Restoring data is a lot of work and a lot more expensive than dumping it with arangodump. While for dumping the data it is sufficient to iterate over all the documents in a collection and stream them, restoring requires checking for conflicts in primary and secondary indexes, and then requires writing the documents plus all their index entries. 
 
For my particular test case (restoring a single collection, 9.5 million documents with 7 attributes each, no seconary indexes), the PR cuts the restore speed in half.
For other use cases, the mileage may vary, but I would expect at least some improvement from it.

I will update this ticket once the above PR gets merged to devel, so it can be tried out in a nightly should there be demand. I will also update the ticket in case the PR will be backported to 3.8, which may be even more useful as the expected GA date for 3.8 is in Q2.

**Comment by admtech** [comment]
An excellent development. I look forward to trying it out in 3.9 devel. Thanks for the effort and support. This is exactly how it should be, you guys did a great job here. Greetings to Cologne from Limburg a.d. Lahn :-)

**Comment by jsteemann** [comment]
@admtech : Regards back to Limburg!

There is now also a backport of the arangorestore improvements to 3.8: https://github.com/arangodb/arangodb/pull/14020
Both PRs (the one for devel and the one for 3.8) have not been reviewed yet, so the changes are not yet available in any nightly or release. I will update here once something gets merged.
In any case, the changes will require using ArangoDB 3.8 or higher, because they rely on the new dump format that is made available only from 3.8 onwards (and this requires a server running 3.8 too). Mentioning it here to manage expectations.

**Comment by admtech** [comment]
Great, thanks for the info. As soon as it is available, we will test it.

**Comment by jsteemann** [comment]
@admtech : the latest change I made has been merged into 3.8 as well now.
It can be tried by building `3.8` from source or by using a nightly 3.8 build that includes the change (i.e. that has a build date >= now).
Nightly 3.8 builds can be found here: https://download.arangodb.com/nightly/3.8/index.html

Please note that nightlies are built only every few days, so it may take some time for the change to appear in a nightly. Note that this is an unreleased 3.8 version which should only be used for testing and nothing else yet.

With a bit of luck the changes will help speeding up the restore a bit, but to manage expectations, I don't expect a 10x speedup or such, rather something like cutting the restore time by 30 to 50%.

If the majority of the time is spent on writing data into the indexes, then my changes will not help so much. They will help mostly if the majority of time is spent waiting on server responses and arangorestore sitting idle meanwhile.

**Comment by admtech** [comment]
Hi,
so I tested the nightly "arangodb3-3.8.0-0.2.x86_64.rpm" today on our test server (Fedora 33, AMD 12 cores, 34 GB Ram, SSD).

The result is more than convincing.

{restore} Processed 14 collection(s) in 313.206299 s, read 2784412168 byte(s) from datafiles, sent 275 data batch(es) of 2784412168 byte(s) total size

We are now at very good times on the import. Currently the system needs 5 to 6 minutes for the restore. 

Here is a small list of our current results:

1 run:
Dump: 2.81 GB DB, 12 Thread, Batchsize 10 MB = 1,23 minutes
Restore: 2.81 GB DB, 6 Thread, Batchsize 10 MB = 6,83 minutes
Restore: 2.81 GB DB, 12 Thread, Batchsize 10 MB = 5,18 minutes
 
2 run:

Dump: 2.81 GB DB, 12 Thread, Batchsize 10 MB = 1,26 minutes
Restore: 2.81 GB DB, 12 Thread, Batchsize 10 MB = 5,12 minutes
Restore: 2.81 GB DB, 12 Thread, ohne batchsize = 5,20 minutes
Restore: 2.81 GB DB, 12 Thread, Batchsize 100 MB = 5,94 minutes

Cross-check version 3.8:

Dump: 2.81 GB DB, 12 Thread, Batchsize 10 MB, envelope false = 1,23 minutes
Restore: 2.81 GB DB, 12 Thread, Batchsize 10 MB, envelope false  = 16,11 minutes

In general, 3.8 seems to be a bit faster than version 3.7 when restoring.

Cross-check version 3.7:

Restore: 2.72 GB DB, 12 Thread, Batchsize 10 MB,  envelope false  = 21,41 minutes

This is an increase of 68% (version 3.8 envelope false) and 76.5% (version 3.7) :-)

Thanks @jsteemann for the good work and fast integration. The journey from "case closed" to import improvement of 50-70% was worth it from our point of view. That's where we like to put time into testing. Thanks to everyone at ArangoDB for making this possible. 

P.S.: maybe you should consider converting the seconds values in the output of arangodump and arangorestore into a directly readable format (h:m:s). Converting the seconds is a hassle every time. Similar is true for the byte values. ;-)

Keep up the good work

**Comment by jsteemann** [comment]
Hi @admtech, thanks for the feedback! 
The performance improvements look indeed encouraging. I was expecting less from it :sweat_smile: 
Regarding your suggestion for better readable outputs (times, bytes) of the client tools, I have addressed this partly via https://github.com/arangodb/arangodb/pull/14081.
This should improve legibility of the size values (bytes, MB, GB, TB) but not yet durations due to lack of time. We may fix this eventually, but I can't guarantee it.
Until then I can recommend prefixing the dump/restoring commands with the `time` builtin, which will automatically print the elapsed time the command has taken.
